### PR TITLE
Fix the NUMA detection for the smcuda BTL.

### PR DESCRIPTION
**Pull request by bosilca**
Code cleanup initiated by a discussion on Slack. If we fail to identify the process NUMA node, or if the process is associated with more than one, print a warning to inform the user about the potential performance implication.

### CODE REVIEW

**Comment by bosilca** [comment]
It all started from a discussion on Slack about a warning in the smcuda BTL pointed out by @wenduwan.

File: opal/mca/btl/smcuda/btl_smcuda.c [hunk 1]
```diff
@@ -247,25 +246,22 @@ static int smcuda_btl_first_time_init(mca_btl_smcuda_t *smcuda_btl, int32_t my_s
...
-            mca_btl_smcuda_component.mem_node = my_mem_node = -1;
-        } else {
+        if (NULL != loc) {
             /* get our NUMA location */
             mynuma = opal_hwloc_base_get_location(loc, HWLOC_OBJ_NODE, 0);
             if (NULL == mynuma || NULL != strchr(mynuma, ',') || NULL != strchr(mynuma, '-')) {
                 /* we either have no idea what NUMA we are on, or we
                  * are on multiple NUMA nodes */
-                mca_btl_smcuda_component.mem_node = my_mem_node = -1;
+                mca_btl_smcuda_component.mem_node = -1;
```
Comment (wenduwan) [inline opal/mca/btl/smcuda/btl_smcuda.c:261]: Now that we initialized `mca_btl_smcuda_component.mem_node = -1;` we can remove this branch.

**Review by wenduwan — COMMENTED** [review]
Looks good to me. Just 1 minor comment.

File: opal/mca/btl/smcuda/btl_smcuda.c (hunk 1 above)
Comment (bosilca) [reply opal/mca/btl/smcuda/btl_smcuda.c:261]: Yeah but nay. The conditional seemed too complicated to just remove the branch, so I decided to keep the true branch and just redo something meaningless (aka. the initialization).

File: opal/mca/btl/smcuda/btl_smcuda.c [hunk 2]
```diff
@@ -297,27 +293,35 @@ static int smcuda_btl_first_time_init(mca_btl_smcuda_t *smcuda_btl, int32_t my_s
...
              */
             if (1 == n_bound) {
-                mca_btl_smcuda_component.mem_node = my_mem_node = numa;
-            } else {
-                mca_btl_smcuda_component.mem_node = my_mem_node = -1;
+                mca_btl_smcuda_component.mem_node = numa;
             }
         }
     }
+    /* sanety check: do we have the NUMA node info ? */
```
Comment (bosilca) [inline opal/mca/btl/smcuda/btl_smcuda.c:300]:
```suggestion
    /* sanity check: do we have the NUMA node info ? */
```

**Comment by wenduwan** [comment]
I tried this patch but still see the warning
```
btl_smcuda.c: In function ‘smcuda_btl_first_time_init’:
btl_smcuda.c:280:27: warning: ‘i’ may be used uninitialized in this function [-Wmaybe-uninitialized]
  280 |             for (w = 0; w < i; w++) {
      |                         ~~^~~
  CCLD     mca_btl_smcuda.la
```

**Comment by wenduwan** [comment]
@bosilca Thank you. The warning is fixed.

Now it revealed something else. `int i` is not useful at all. On L427 I see
```
 427     i = opal_free_list_init(&mca_btl_smcuda_component.sm_frags_eager, length, opal_cache_line_size,
 428                             OBJ_CLASS(mca_btl_smcuda_frag1_t), length_payload, opal_cache_line_size,
 429                             mca_btl_smcuda_component.sm_free_list_num,
 430                             mca_btl_smcuda_component.sm_free_list_max,
 431                             mca_btl_smcuda_component.sm_free_list_inc,
 432                             mca_btl_smcuda_component.sm_mpool, 0, NULL, NULL, NULL);
 433     if (OPAL_SUCCESS != i)
 434         return i;
```
`i` is used instead of `rc`.

**Review by wenduwan — APPROVED** [review]

**Comment by wenduwan** [comment]
Thanks @bosilca . I will run some additional tests in the background(I don't expect any issue).

Feel free to merge otherwise.

**Pull request merged**